File: kernel/net/udp.c
/**
 * @file udp.c
 * @brief UDP protocol implementation
 */

#include "../include/net/udp.h"
#include "../include/net/ip.h"
#include "../include/kprintf.h"
#include "../include/debug.h"
#include "../include/mm/heap.h"
#include "../include/string.h"

/**
 * Send UDP packet
 */
error_code_t udp_send(uint32_t dest_ip, uint16_t dest_port, uint16_t src_port, void* data, size_t len) {
    if (!data || len == 0) {
        return ERR_INVALID_ARG;
    }
    
    // Allocate UDP packet
    size_t packet_size = sizeof(udp_header_t) + len;
    udp_header_t* packet = (udp_header_t*)kmalloc(packet_size);
    if (!packet) {
        return ERR_OUT_OF_MEMORY;
    }
    
    // Build UDP header
    packet->src_port = __builtin_bswap16(src_port);
    packet->dest_port = __builtin_bswap16(dest_port);
    packet->length = __builtin_bswap16(packet_size);
    packet->checksum = 0;  // Optional for UDP
    
    // Copy data
    memcpy(packet->data, data, len);
    
    // Send via IP
    error_code_t err = ip_send(dest_ip, IP_PROTOCOL_UDP, packet, packet_size);
    
    kfree(packet);
    return err;
}

/**
 * Receive UDP packet
 * TODO: Fix to match new IP layer signature
 */
error_code_t udp_receive(void* buffer, size_t* len, uint32_t* src_ip, uint16_t* src_port, uint16_t* dest_port) {
    if (!buffer || !len || !src_ip || !src_port || !dest_port) {
        return ERR_INVALID_ARG;
    }

    // TODO: Implement proper UDP receive with new IP layer API
    (void)buffer;
    (void)len;
    (void)src_ip;
    (void)src_port;
    (void)dest_port;
    return ERR_NOT_SUPPORTED;
}

File: services/network/src/udp.rs
//! UDP protocol implementation

/// UDP header structure
#[repr(C, packed)]
pub struct UdpHeader {
    pub src_port: u16,
    pub dest_port: u16,
    pub length: u16,
    pub checksum: u16,
    pub data: [u8; 0],  // Variable length data
}

/// Send UDP packet
pub fn udp_send(dest_ip: u32, dest_port: u16, src_port: u16, data: &[u8]) -> Result<(), ()> {
    // Build UDP header
    let mut udp_header = UdpHeader {
        src_port,
        dest_port,
        length: (8 + data.len()) as u16,
        checksum: 0, // Checksum would be calculated here
        data: [],
    };
    
    // Build packet
    let mut packet = [0u8; 1500];
    unsafe {
        core::ptr::copy_nonoverlapping(&udp_header as *const _ as *const u8, packet.as_mut_ptr(), 8);
    }
    let data_len = data.len().min(1492);
    packet[8..8+data_len].copy_from_slice(&data[0..data_len]);
    
    // Calculate checksum (simplified - would include pseudo-header)
    // For now, skip checksum calculation
    
    // Send via IP layer
    use crate::ip::ip_send;
    ip_send(dest_ip, crate::ip::IP_PROTOCOL_UDP, &packet[0..8+data_len])
}

/// Receive UDP packet
pub fn udp_receive(buffer: &mut [u8]) -> Result<(usize, u32, u16, u16), ()> {
    // Receive from IP layer
    use crate::ip::ip_receive;
    let mut ip_buffer = [0u8; 1500];
    match ip_receive(&mut ip_buffer) {
        Ok((len, src_ip, protocol)) => {
            if protocol == crate::ip::IP_PROTOCOL_UDP && len >= 8 {
                // Parse UDP header
                let udp_header = unsafe {
                    &*(ip_buffer.as_ptr() as *const UdpHeader)
                };
                
                // Copy data to buffer
                let data_len = (len - 8).min(buffer.len());
                buffer[0..data_len].copy_from_slice(&ip_buffer[8..8+data_len]);
                
                // Return data length, source IP, source port, dest port
                Ok((data_len, src_ip, udp_header.src_port, udp_header.dest_port))
            } else {
                Err(())
            }
        }
        Err(_) => Err(())
    }
}

